### DISCUSSION TITLE
Is Clearscript right way to go?

### DISCUSSION THREAD

**Discussion by koculu** [ClearFoundry/ClearScript#313] [General]
I have written a JS engine in pure C# which supports multithreading. 
Performance and memory consumption is far more better than Clearscript and good for server-side rendering.
https://github.com/koculu/topaz
Please take a look. I posted here for those who might be interested in.
It seems like marshaling to C++ is not a good way for .NET applications.

**Comment by ClearScriptLib** [comment]
Hi @koculu,

Your project, Topaz, looks fantastic! Thanks for releasing it on GitHub!

There's no question that V8 heavily favors script execution speed over startup time and memory efficiency, and that the boundary between the managed host and the V8 environment is expensive. If your application depends on chatty host-script interaction, V8 is probably not for you, especially if your scripts are small and not computationally intensive.

As for concurrency, the usual way to achieve it in JavaScript is by using multiple engine instances that communicate via message passing. [Web Workers](https://html.spec.whatwg.org/multipage/workers.html#workers) are an example of this approach. We've posted a [sample](https://github.com/microsoft/ClearScriptWorkerSample) that demonstrates a similar facility implemented with ClearScript.

By the way, we modified the Topaz benchmark to print results for Topaz, Jint, and ClearScript. On our test machine, it produces the following output:

```
Topaz: 220 ms
Jint: 233 ms
ClearScript: 48 ms
```

If we raise the iteration count from 100,000 to 10,000,000, the results are as follows:

```
Topaz: 6818 ms
Jint: 8280 ms
ClearScript: 63 ms
```

The bottom line is that there are always tradeoffs 😉

As this topic doesn't refer to an actionable ClearScript issue, we've converted it to a discussion. Please feel free to send additional thoughts or findings our way.

Thanks!

**Reply by koculu** [reply]
Thank you for clarification! It is nice to know use cases and performance characteristics like this. I already saw huge difference between .NET and Topaz but didn't check Clearscript.
I have to think about how to improve efficiency of iterations in Topaz :) 
In my business case 10M iterations is not a need.

Thank you.

**Reply by koculu** [reply]
After a couple of optimizations and variable local caching I succeeded to improve
i = 0..10M i+i loop in single thread and parallel execution.

```
Single Thread Topaz Loop 10M: 4987 ms
Parallel Loop 10M: 1109 ms
```

**Comment by koculu** [comment]
Parallel ForEach improves the efficiency of for loops. Following code works now on Topaz.
So in general the loop speed is acceptable for many real world use cases. FYI
```c#
var engine = new TopazEngine();
engine.AddType(typeof(Console), "Console");
topazEngine.AddType(typeof(Parallel), "Parallel");
engine.ExecuteScript(@"
var sharedVariable = 0
function f1(i) {
    sharedVariable = i
}
Parallel.For(0, 100000 , f1)
Console.WriteLine(`Final value: {sharedVariable}`);
");
```

**Comment by koculu** [comment]
Latest benchmark results for script engines:

1M for loop with no host interaction:
```` ini
|           Method |       Mean |      Error |     StdDev |
|----------------- |-----------:|-----------:|-----------:|
|      RunV8Engine |   6.245 ms |  0.0423 ms |  0.0396 ms |
| RunParallelTopaz | 125.584 ms |  2.4922 ms |  5.2022 ms |
|         RunTopaz | 523.444 ms | 10.3078 ms |  9.6419 ms |
|          RunJint | 882.752 ms | 13.7271 ms | 10.7172 ms |
````

1M manipulations on arbitrary .NET objects
```` ini
|   Method     |     Mean    |   Error   |  StdDev   |
|------------- |------------:|----------:|----------:|
| RunTopaz     | 792.0 ms    | 9.74 ms   | 9.11 ms   |
| RunJint      | 896.6 ms    | 7.78 ms   | 7.28 ms   |
| RunV8Engine  | 30,598.6 ms |    NA     |    NA     |
````

For more benchmarks see: https://github.com/koculu/topaz
